Feature: API to implement alert silencing

venkateshamatam opened the pull request:
# Description
Added functionality to silence alerts for a certain time period. One can set "silence_minutes" in minutes via the API(/api/alerts/silenceAlert) that will prevent notifications from being sent for that alert, while still keeping the alert active behind the scenes.

### Includes:

- New "silence_minutes" column added to Alerts table to store length of silence period
- New "/alerts/silenceAlert" API that accepts alert ID and silence period and updates the Alerts table
- Updated alert notification logic to check if alert is silenced before sending notifications

Fixes #36

# Testing
- Call PUT /alerts/silenceAlert API with an existing alert ID and a silence period in hours:
```
{
  "alert_id": "17471435-3abd-4c06-8f7d-aa3bed0a7a40",
  "silence_minutes": 15
}
```
- Verify entry added to Alerts table with silence_minutes by calling GET api/alerts/{alertId}
- Trigger alert and validate no notification sent out during silence period
- Wait for silence period to expire, trigger alert again and validate notification is sent

# Checklist:
Before marking your pull request as ready for review, complete the following.

- [x] I have self-reviewed this PR.
- [x] I have removed all print-debugging and commented-out code that should not be merged.
- [x] I have added sufficient comments in my code, particularly in hard-to-understand areas.
- [x] I have formatted the code, if applicable. For Go, I have run `goimports -w .`.

AndrewHess reviewed (commented):
This is looking good. A few things:
1. Instead of having the silence period in hours as a float, I think it would make more sense to have it in minutes as an integer.
2. Can you change the `silencePeriod` to something that indicates its units (e.g., `silenceMinutes`)?
3. if someone silenced an alert and then wanted to unsilence it, they could call the API and specify a silence time of 0, right?

venkateshamatam commented on pkg/alerts/alertsHandler/notificationHandler.go:
```diff
@@ -143,6 +151,26 @@ func sendWebhooks(webhookUrl, subject, message string) error {
+
+	if err != nil {
+		return false, err
+	}
```
@AndrewHess I agree, I think it's better to fail-safe in the context of an alert system. I've updated the function to return true in case of an error during the silence period.

venkateshamatam commented:
> This is looking good. A few things:
> 
> 1. Instead of having the silence period in hours as a float, I think it would make more sense to have it in minutes as an integer.
> 2. Can you change the `silencePeriod` to something that indicates its units (e.g., `silenceMinutes`)?
> 3. if someone silenced an alert and then wanted to unsilence it, they could call the API and specify a silence time of 0, right?

@AndrewHess, I've committed the changes - 

1. I've updated it to store it as an unsigned integer.
2. I have changed the name to silenceMinutes and json to silence_minutes and the logic now treats SilenceMinutes as minutes and it is not possible to pass non-integer values.
3. Yes, is is possible to unsilence the alert by setting the value to 0 and it gets the latest value (i.e 0 if you unsilence it) to see if silencePeriod is over before sending an alert.

venkateshamatam commented:
@AndrewHess, Could you please have a look at this PR and let me know if any changes need to be made?

venkateshamatam commented:
@AndrewHess,This PR is ready for review. Could you please have a look at this PR and let me know if any changes need to be made?

nkunal commented:
@sunitakawane can you review this PR please

AndrewHess reviewed (approved):
Looks good! (Sorry for the late review)
